File: crous-core/src/text.rs
//! Human-readable Crous text format: parser and pretty-printer.
//!
//! # Crous Text Syntax (summary)
//!
//! The Crous textual notation is a unique, deterministic syntax that maps
//! 1:1 to the binary format. It is NOT a JSON clone — it has its own rules:
//!
//! - Objects use `{ key: value; key2: value2; }` with `;` as mandatory terminator.
//! - Arrays use `[ value, value, value ]` with `,` separator.
//! - Strings are double-quoted: `"hello world"`.
//! - Binary data uses `b64#<base64>;` marker.
//! - Integers: unsigned are bare digits, signed use `+` or `-` prefix.
//! - Floats use decimal point: `3.14`, `-2.718`.
//! - Null is `null`, booleans are `true`/`false`.
//! - Optional type annotations: `42::u32`, `"hello"::str`.
//! - Comments: `// line comment` and `/* block comment */`.
//!
//! # ABNF Grammar
//!
//! ```abnf
//! document     = value
//! value        = null / boolean / integer / float / string / bytes
//!              / array / object
//! null         = "null"
//! boolean      = "true" / "false"
//! integer      = [sign] 1*DIGIT [type-ann]
//! float        = [sign] 1*DIGIT "." 1*DIGIT [exponent] [type-ann]
//! exponent     = ("e" / "E") [sign] 1*DIGIT
//! sign         = "+" / "-"
//! string       = DQUOTE *char DQUOTE [type-ann]
//! bytes        = "b64#" base64-data ";"
//! base64-data  = *( ALPHA / DIGIT / "+" / "/" / "=" )
//! array        = "[" [value *("," value)] "]"
//! object       = "{" *field "}"
//! field        = key ":" value ";"
//! key          = identifier / string
//! identifier   = (ALPHA / "_") *(ALPHA / DIGIT / "_")
//! type-ann     = "::" type-name
//! type-name    = identifier
//! comment      = line-comment / block-comment
//! line-comment = "//" *(%x20-7E) LF
//! block-comment= "/*" *(comment-char) "*/"
//! ```

use crate::error::{CrousError, Result};
use crate::value::Value;
use base64::Engine;

// ---------------------------------------------------------------------------
// Parser
// ---------------------------------------------------------------------------

/// Parse a Crous text document into a `Value`.
///
/// ```
/// use crous_core::text::parse;
/// use crous_core::Value;
///
/// let v = parse(r#"{ name: "Alice"; age: 30; }"#).unwrap();
/// assert_eq!(
///     v,
///     Value::Object(vec![
///         ("name".into(), Value::Str("Alice".into())),
///         ("age".into(), Value::UInt(30)),
///     ])
/// );
/// ```
pub fn parse(input: &str) -> Result<Value> {
    let mut parser = Parser::new(input);
    let value = parser.parse_value()?;
    parser.skip_whitespace_and_comments();
    if parser.peek() == Some(';') {
        parser.advance();
        parser.skip_whitespace_and_comments();
    }
    if parser.pos != input.len() {
        return Err(parser.error("unexpected trailing content"));
    }
    Ok(value)
}

struct Parser<'a> {
    input: &'a str,
    pos: usize,
    line: usize,
    col: usize,
}

impl<'a> Parser<'a> {
    fn new(input: &'a str) -> Self {
        Self {
            input,
            pos: 0,
            line: 1,
            col: 1,
        }
    }

    fn peek(&self) -> Option<char> {
        self.input[self.pos..].chars().next()
    }

    fn advance(&mut self) -> Option<char> {
        let ch = self.peek()?;
        self.pos += ch.len_utf8();
        if ch == '\n' {
            self.line += 1;
            self.col = 1;
        } else {
            self.col += 1;
        }
        Some(ch)
    }

    fn remaining(&self) -> &'a str {
        &self.input[self.pos..]
    }

    fn error(&self, msg: impl Into<String>) -> CrousError {
        CrousError::ParseError {
            line: self.line,
            col: self.col,
            message: msg.into(),
        }
    }

    fn skip_whitespace_and_comments(&mut self) {
        loop {
            // Skip whitespace.
            while let Some(ch) = self.peek() {
                if ch.is_whitespace() {
                    self.advance();
                } else {
                    break;
                }
            }
            // Skip line comments.
            if self.remaining().starts_with("//") {
                while let Some(ch) = self.advance() {
                    if ch == '\n' {
                        break;
                    }
                }
                continue;
            }
            // Skip block comments.
            if self.remaining().starts_with("/*") {
                self.advance(); // '/'
                self.advance(); // '*'
                let mut depth = 1;
                while depth > 0 {
                    match self.advance() {
                        Some('*') if self.peek() == Some('/') => {
                            self.advance();
                            depth -= 1;
                        }
                        Some('/') if self.peek() == Some('*') => {
                            self.advance();
                            depth += 1;
                        }
                        Some(_) => {}
                        None => break,
                    }
                }
                continue;
            }
            break;
        }
    }

    fn expect_char(&mut self, expected: char) -> Result<()> {
        self.skip_whitespace_and_comments();
        match self.advance() {
            Some(ch) if ch == expected => Ok(()),
            Some(ch) => Err(self.error(format!("expected '{expected}', got '{ch}'"))),
            None => Err(self.error(format!("expected '{expected}', got EOF"))),
        }
    }

    fn parse_value(&mut self) -> Result<Value> {
        self.skip_whitespace_and_comments();

        match self.peek() {
            None => Err(self.error("unexpected end of input")),
            Some('{') => self.parse_object(),
            Some('[') => self.parse_array(),
            Some('"') => self.parse_string_value(),
            Some('b') if self.remaining().starts_with("b64#") => self.parse_bytes(),
            Some('t') if self.remaining().starts_with("true") => self.parse_true(),
            Some('f') if self.remaining().starts_with("false") => self.parse_false(),
            Some('n') if self.remaining().starts_with("null") => self.parse_null(),
            Some('i') if self.remaining().starts_with("inf") => self.parse_inf(false),
            Some('N') if self.remaining().starts_with("NaN") => self.parse_nan(),
            Some(ch) if ch == '-' || ch == '+' || ch.is_ascii_digit() => {
                // Check for "-inf"
                if ch == '-' && self.remaining().starts_with("-inf") {
                    self.parse_inf(true)
                } else {
                    self.parse_number()
                }
            }
            Some(ch) => Err(self.error(format!("unexpected character: '{ch}'"))),
        }
    }

    fn parse_null(&mut self) -> Result<Value> {
        for _ in 0..4 {
            self.advance();
        }
        self.skip_type_annotation();
        Ok(Value::Null)
    }

    fn parse_inf(&mut self, negative: bool) -> Result<Value> {
        if negative {
            // skip "-inf"
            for _ in 0..4 {
                self.advance();
            }
            self.skip_type_annotation();
            Ok(Value::Float(f64::NEG_INFINITY))
        } else {
            // skip "inf"
            for _ in 0..3 {
                self.advance();
            }
            self.skip_type_annotation();
            Ok(Value::Float(f64::INFINITY))
        }
    }

    fn parse_nan(&mut self) -> Result<Value> {
        for _ in 0..3 {
            self.advance();
        }
        self.skip_type_annotation();
        Ok(Value::Float(f64::NAN))
    }

    fn parse_true(&mut self) -> Result<Value> {
        for _ in 0..4 {
            self.advance();
        }
        self.skip_type_annotation();
        Ok(Value::Bool(true))
    }

    fn parse_false(&mut self) -> Result<Value> {
        for _ in 0..5 {
            self.advance();
        }
        self.skip_type_annotation();
        Ok(Value::Bool(false))
    }

    fn parse_number(&mut self) -> Result<Value> {
        let start = self.pos;
        let mut is_negative = false;
        let mut is_float = false;

        if self.peek() == Some('-') {
            is_negative = true;
            self.advance();
        } else if self.peek() == Some('+') {
            self.advance();
        }

        while let Some(ch) = self.peek() {
            if ch.is_ascii_digit() {
                self.advance();
            } else if ch == '.' {
                is_float = true;
                self.advance();
            } else if ch == 'e' || ch == 'E' {
                is_float = true;
                self.advance();
                if self.peek() == Some('+') || self.peek() == Some('-') {
                    self.advance();
                }
            } else {
                break;
            }
        }

        let num_str = &self.input[start..self.pos];
        self.skip_type_annotation();

        if is_float {
            let f: f64 = num_str
                .parse()
                .map_err(|_| self.error(format!("invalid float: {num_str}")))?;
            Ok(Value::Float(f))
        } else if is_negative {
            let i: i64 = num_str
                .parse()
                .map_err(|_| self.error(format!("invalid integer: {num_str}")))?;
            Ok(Value::Int(i))
        } else {
            let u: u64 = num_str
                .parse()
                .map_err(|_| self.error(format!("invalid integer: {num_str}")))?;
            Ok(Value::UInt(u))
        }
    }

    fn parse_string_value(&mut self) -> Result<Value> {
        let s = self.parse_quoted_string()?;
        self.skip_type_annotation();
        Ok(Value::Str(s))
    }

    fn parse_quoted_string(&mut self) -> Result<String> {
        self.expect_char('"')?;
        let mut s = String::new();
        loop {
            match self.advance() {
                Some('"') => break,
                Some('\\') => match self.advance() {
                    Some('n') => s.push('\n'),
                    Some('t') => s.push('\t'),
                    Some('r') => s.push('\r'),
                    Some('\\') => s.push('\\'),
                    Some('"') => s.push('"'),
                    Some(ch) => {
                        s.push('\\');
                        s.push(ch);
                    }
                    None => return Err(self.error("unterminated string escape")),
                },
                Some(ch) => s.push(ch),
                None => return Err(self.error("unterminated string")),
            }
        }
        Ok(s)
    }

    fn parse_bytes(&mut self) -> Result<Value> {
        // Consume "b64#"
        for _ in 0..4 {
            self.advance();
        }
        let start = self.pos;
        // Read only base64 payload characters; do not consume the next delimiter.
        // This allows bytes values in objects (delimiter ';') and arrays (delimiter ',').
        while let Some(ch) = self.peek() {
            if ch.is_ascii_alphanumeric() || ch == '+' || ch == '/' || ch == '=' {
                self.advance();
            } else {
                break;
            }
        }
        let b64_str = &self.input[start..self.pos];

        if b64_str.is_empty() {
            return Err(self.error("empty base64 payload"));
        }

        let bytes = base64::engine::general_purpose::STANDARD
            .decode(b64_str.trim())
            .map_err(|e| self.error(format!("invalid base64: {e}")))?;
        Ok(Value::Bytes(bytes))
    }

    fn parse_array(&mut self) -> Result<Value> {
        self.expect_char('[')?;
        let mut items = Vec::new();

        loop {
            self.skip_whitespace_and_comments();
            if self.peek() == Some(']') {
                self.advance();
                break;
            }
            items.push(self.parse_value()?);
            self.skip_whitespace_and_comments();
            if self.peek() == Some(',') || self.peek() == Some(';') {
                self.advance();
            }
        }

        Ok(Value::Array(items))
    }

    fn parse_object(&mut self) -> Result<Value> {
        self.expect_char('{')?;
        let mut entries = Vec::new();

        loop {
            self.skip_whitespace_and_comments();
            if self.peek() == Some('}') {
                self.advance();
                break;
            }

            // Parse key: either an identifier or a quoted string.
            let key = self.parse_key()?;
            self.expect_char(':')?;
            let value = self.parse_value()?;
            self.expect_char(';')?;

            entries.push((key, value));
        }

        Ok(Value::Object(entries))
    }

    fn parse_key(&mut self) -> Result<String> {
        self.skip_whitespace_and_comments();
        if self.peek() == Some('"') {
            self.parse_quoted_string()
        } else {
            self.parse_identifier()
        }
    }

    fn parse_identifier(&mut self) -> Result<String> {
        let start = self.pos;
        match self.peek() {
            Some(ch) if ch.is_alphabetic() || ch == '_' => {
                self.advance();
            }
            _ => return Err(self.error("expected identifier")),
        }
        while let Some(ch) = self.peek() {
            if ch.is_alphanumeric() || ch == '_' {
                self.advance();
            } else {
                break;
            }
        }
        Ok(self.input[start..self.pos].to_string())
    }

    /// Skip optional type annotations like `::u32`, `::str`.
    fn skip_type_annotation(&mut self) {
        if self.remaining().starts_with("::") {
            self.advance(); // ':'
            self.advance(); // ':'
            while let Some(ch) = self.peek() {
                if ch.is_alphanumeric() || ch == '_' {
                    self.advance();
                } else {
                    break;
                }
            }
        }
    }
}

// ---------------------------------------------------------------------------
// Pretty-printer
// ---------------------------------------------------------------------------

/// Pretty-print a `Value` in canonical Crous text notation.
///
/// The output is deterministic: the same `Value` always produces the same text.
///
/// ```
/// use crous_core::text::pretty_print;
/// use crous_core::Value;
///
/// let v = Value::Object(vec![
///     ("name".into(), Value::Str("Alice".into())),
///     ("age".into(), Value::UInt(30)),
/// ]);
/// let text = pretty_print(&v, 0);
/// assert!(text.contains("name: \"Alice\";"));
/// assert!(text.contains("age: 30;"));
/// ```
pub fn pretty_print(value: &Value, indent: usize) -> String {
    let mut out = String::new();
    write_value(&mut out, value, indent, 0);
    out
}

fn write_value(out: &mut String, value: &Value, indent_size: usize, depth: usize) {
    let indent = " ".repeat(indent_size * depth);
    let inner_indent = " ".repeat(indent_size * (depth + 1));

    match value {
        Value::Null => out.push_str("null"),
        Value::Bool(b) => out.push_str(if *b { "true" } else { "false" }),
        Value::UInt(n) => out.push_str(&n.to_string()),
        Value::Int(n) => {
            // Int(0) pretty-prints as "0" which on reparse becomes UInt(0).
            // Emit a negative-zero form to preserve the Int type for roundtrip.
            if *n == 0 {
                out.push_str("-0");
            } else {
                out.push_str(&n.to_string());
            }
        }
        Value::Float(f) => {
            if f.is_nan() {
                out.push_str("NaN");
            } else if f.is_infinite() {
                if f.is_sign_negative() {
                    out.push_str("-inf");
                } else {
                    out.push_str("inf");
                }
            } else {
                // Ensure float always has a decimal point for deterministic output.
                let s = format!("{f}");
                if s.contains('.') || s.contains('e') || s.contains('E') {
                    out.push_str(&s);
                } else {
                    out.push_str(&format!("{f}.0"));
                }
            }
        }
        Value::Str(s) => {
            out.push('"');
            for ch in s.chars() {
                match ch {
                    '"' => out.push_str("\\\""),
                    '\\' => out.push_str("\\\\"),
                    '\n' => out.push_str("\\n"),
                    '\r' => out.push_str("\\r"),
                    '\t' => out.push_str("\\t"),
                    c => out.push(c),
                }
            }
            out.push('"');
        }
        Value::Bytes(b) => {
            out.push_str("b64#");
            out.push_str(&base64::engine::general_purpose::STANDARD.encode(b));
        }
        Value::Array(items) => {
            if items.is_empty() {
                out.push_str("[]");
            } else if is_simple_array(items) {
                // Inline for simple arrays.
                out.push('[');
                for (i, item) in items.iter().enumerate() {
                    if i > 0 {
                        out.push_str(", ");
                    }
                    write_value(out, item, indent_size, depth);
                }
                out.push(']');
            } else {
                out.push_str("[\n");
                for (i, item) in items.iter().enumerate() {
                    out.push_str(&inner_indent);
                    write_value(out, item, indent_size, depth + 1);
                    if i < items.len() - 1 {
                        out.push(',');
                    }
                    out.push('\n');
                }
                out.push_str(&indent);
                out.push(']');
            }
        }
        Value::Object(entries) => {
            if entries.is_empty() {
                out.push_str("{}");
            } else {
                out.push_str("{\n");
                for (key, val) in entries {
                    out.push_str(&inner_indent);
                    if is_valid_identifier(key) {
                        out.push_str(key);
                    } else {
                        out.push('"');
                        out.push_str(key);
                        out.push('"');
                    }
                    out.push_str(": ");
                    write_value(out, val, indent_size, depth + 1);
                    out.push_str(";\n");
                }
                out.push_str(&indent);
                out.push('}');
            }
        }
    }
}

/// Check if an array contains only simple scalar values (no nesting).
fn is_simple_array(items: &[Value]) -> bool {
    items.len() <= 8
        && items.iter().all(|v| {
            matches!(
                v,
                Value::Null
                    | Value::Bool(_)
                    | Value::UInt(_)
                    | Value::Int(_)
                    | Value::Float(_)
                    | Value::Str(_)
            )
        })
}

/// Check if a string is a valid unquoted identifier.
fn is_valid_identifier(s: &str) -> bool {
    let mut chars = s.chars();
    match chars.next() {
        Some(c) if c.is_alphabetic() || c == '_' => {}
        _ => return false,
    }
    chars.all(|c| c.is_alphanumeric() || c == '_')
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_null() {
        assert_eq!(parse("null").unwrap(), Value::Null);
    }

    #[test]
    fn parse_bool() {
        assert_eq!(parse("true").unwrap(), Value::Bool(true));
        assert_eq!(parse("false").unwrap(), Value::Bool(false));
    }

    #[test]
    fn parse_uint() {
        assert_eq!(parse("42").unwrap(), Value::UInt(42));
        assert_eq!(parse("0").unwrap(), Value::UInt(0));
    }

    #[test]
    fn parse_int() {
        assert_eq!(parse("-1").unwrap(), Value::Int(-1));
        assert_eq!(parse("-42").unwrap(), Value::Int(-42));
    }

    #[test]
    fn parse_float() {
        assert_eq!(parse("3.125").unwrap(), Value::Float(3.125));
        assert_eq!(parse("-2.5").unwrap(), Value::Float(-2.5));
    }

    #[test]
    fn parse_string() {
        assert_eq!(parse(r#""hello""#).unwrap(), Value::Str("hello".into()));
        assert_eq!(
            parse(r#""with \"quotes\"""#).unwrap(),
            Value::Str("with \"quotes\"".into())
        );
    }

    #[test]
    fn parse_bytes() {
        let v = parse("b64#AQID;").unwrap();
        assert_eq!(v, Value::Bytes(vec![1, 2, 3]));
    }

    #[test]
    fn parse_array() {
        let v = parse("[1, 2, 3]").unwrap();
        assert_eq!(
            v,
            Value::Array(vec![Value::UInt(1), Value::UInt(2), Value::UInt(3)])
        );
    }

    #[test]
    fn parse_object() {
        let v = parse(r#"{ name: "Alice"; age: 30; }"#).unwrap();
        assert_eq!(
            v,
            Value::Object(vec![
                ("name".into(), Value::Str("Alice".into())),
                ("age".into(), Value::UInt(30)),
            ])
        );
    }

    #[test]
    fn parse_nested() {
        let input = r#"{
            users: [
                { name: "Bob"; scores: [100, 95, 87]; }
            ];
            count: 1;
        }"#;
        let v = parse(input).unwrap();
        let expected = Value::Object(vec![
            (
                "users".into(),
                Value::Array(vec![Value::Object(vec![
                    ("name".into(), Value::Str("Bob".into())),
                    (
                        "scores".into(),
                        Value::Array(vec![Value::UInt(100), Value::UInt(95), Value::UInt(87)]),
                    ),
                ])]),
            ),
            ("count".into(), Value::UInt(1)),
        ]);
        assert_eq!(v, expected);
    }

    #[test]
    fn parse_comments() {
        let input = r#"{
            // This is a comment
            name: "Alice"; /* inline comment */
            age: 30;
        }"#;
        let v = parse(input).unwrap();
        assert_eq!(
            v,
            Value::Object(vec![
                ("name".into(), Value::Str("Alice".into())),
                ("age".into(), Value::UInt(30)),
            ])
        );
    }

    #[test]
    fn parse_type_annotation() {
        let v = parse("42::u32").unwrap();
        assert_eq!(v, Value::UInt(42));
    }

    #[test]
    fn parse_rejects_trailing_content() {
        assert!(parse("42 trailing").is_err());
    }

    #[test]
    fn parse_bytes_inside_array() {
        let v = parse("[b64#AQID, b64#BAUG]").unwrap();
        assert_eq!(
            v,
            Value::Array(vec![
                Value::Bytes(vec![1, 2, 3]),
                Value::Bytes(vec![4, 5, 6])
            ])
        );

        let v2 = parse("[b64#AQID; b64#BAUG;]").unwrap();
        assert_eq!(
            v2,
            Value::Array(vec![
                Value::Bytes(vec![1, 2, 3]),
                Value::Bytes(vec![4, 5, 6])
            ])
        );
    }

    #[test]
    fn pretty_print_roundtrip() {
        let original = Value::Object(vec![
            ("name".into(), Value::Str("Alice".into())),
            ("age".into(), Value::UInt(30)),
            (
                "tags".into(),
                Value::Array(vec![Value::Str("admin".into()), Value::Str("user".into())]),
            ),
        ]);
        let text = pretty_print(&original, 4);
        let parsed = parse(&text).unwrap();
        assert_eq!(parsed, original);
    }

    #[test]
    fn pretty_print_bytes() {
        let v = Value::Bytes(vec![0xDE, 0xAD, 0xBE, 0xEF]);
        let text = pretty_print(&v, 0);
        assert!(text.starts_with("b64#"));
        // b64# values no longer include trailing ';' — the terminator is handled
        // by the enclosing object/array syntax.
        let parsed = parse(&text).unwrap();
        assert_eq!(parsed, v);
    }

    #[test]
    fn text_binary_text_roundtrip() {
        // Parse text → encode binary → decode binary → pretty-print text → parse text
        let input = r#"{ name: "Alice"; age: 30; active: true; }"#;
        let val1 = parse(input).unwrap();

        let mut enc = crate::encoder::Encoder::new();
        enc.encode_value(&val1).unwrap();
        let binary = enc.finish().unwrap();

        let mut dec = crate::decoder::Decoder::new(&binary);
        let val2 = dec.decode_next().unwrap().to_owned_value();
        assert_eq!(val1, val2);

        let text2 = pretty_print(&val2, 4);
        let val3 = parse(&text2).unwrap();
        assert_eq!(val1, val3);
    }
}

File: python/surp/text.py
"""Human-readable Surp text format: parser and pretty-printer."""

from __future__ import annotations

import base64

from surp.error import ParseError
from surp.value import Value, ValueType


# ---------------------------------------------------------------------------
# Parser
# ---------------------------------------------------------------------------


def parse(input_text: str) -> Value:
    """Parse a Surp text document into a Value.

    >>> parse('null')
    Value.null()
    >>> parse('42')
    Value.uint(42)
    >>> parse('"hello"')
    Value.str_('hello')
    """
    parser = _Parser(input_text)
    value = parser.parse_value()
    parser.skip_ws()
    return value


class _Parser:
    def __init__(self, text: str):
        self._text = text
        self._pos = 0
        self._line = 1
        self._col = 1

    def _peek(self) -> str | None:
        if self._pos >= len(self._text):
            return None
        return self._text[self._pos]

    def _advance(self) -> str | None:
        ch = self._peek()
        if ch is None:
            return None
        self._pos += 1
        if ch == "\n":
            self._line += 1
            self._col = 1
        else:
            self._col += 1
        return ch

    def _remaining(self) -> str:
        return self._text[self._pos :]

    def _error(self, msg: str) -> ParseError:
        return ParseError(self._line, self._col, msg)

    def skip_ws(self) -> None:
        while True:
            # Skip whitespace
            while self._peek() is not None and self._peek() in " \t\r\n":
                self._advance()

            # Skip line comments
            if self._remaining().startswith("//"):
                while True:
                    ch = self._advance()
                    if ch is None or ch == "\n":
                        break
                continue

            # Skip block comments
            if self._remaining().startswith("/*"):
                self._advance()  # /
                self._advance()  # *
                depth = 1
                while depth > 0:
                    ch = self._advance()
                    if ch is None:
                        break
                    if ch == "*" and self._peek() == "/":
                        self._advance()
                        depth -= 1
                    elif ch == "/" and self._peek() == "*":
                        self._advance()
                        depth += 1
                continue

            break

    def _expect_char(self, expected: str) -> None:
        self.skip_ws()
        ch = self._advance()
        if ch != expected:
            got = repr(ch) if ch else "EOF"
            raise self._error(f"expected '{expected}', got {got}")

    def parse_value(self) -> Value:
        self.skip_ws()
        ch = self._peek()

        if ch is None:
            raise self._error("unexpected end of input")
        if ch == "{":
            return self._parse_object()
        if ch == "[":
            return self._parse_array()
        if ch == '"':
            return self._parse_string_value()
        if self._remaining().startswith("b64#"):
            return self._parse_bytes()
        if self._remaining().startswith("true"):
            return self._parse_true()
        if self._remaining().startswith("false"):
            return self._parse_false()
        if self._remaining().startswith("null"):
            return self._parse_null()
        if ch in "-+0123456789":
            return self._parse_number()

        raise self._error(f"unexpected character: '{ch}'")

    def _parse_null(self) -> Value:
        for _ in range(4):
            self._advance()
        self._skip_type_annotation()
        return Value.null()

    def _parse_true(self) -> Value:
        for _ in range(4):
            self._advance()
        self._skip_type_annotation()
        return Value.bool_(True)

    def _parse_false(self) -> Value:
        for _ in range(5):
            self._advance()
        self._skip_type_annotation()
        return Value.bool_(False)

    def _parse_number(self) -> Value:
        start = self._pos
        is_negative = False
        is_float = False

        if self._peek() == "-":
            is_negative = True
            self._advance()
        elif self._peek() == "+":
            self._advance()

        while self._peek() is not None:
            ch = self._peek()
            if ch is not None and ch.isdigit():
                self._advance()
            elif ch == ".":
                is_float = True
                self._advance()
            elif ch in "eE":
                is_float = True
                self._advance()
                if self._peek() in "+-":
                    self._advance()
            else:
                break

        num_str = self._text[start : self._pos]
        self._skip_type_annotation()

        if is_float:
            return Value.float_(float(num_str))
        elif is_negative:
            return Value.int_(int(num_str))
        else:
            return Value.uint(int(num_str))

    def _parse_string_value(self) -> Value:
        s = self._parse_quoted_string()
        self._skip_type_annotation()
        return Value.str_(s)

    def _parse_quoted_string(self) -> str:
        self._expect_char('"')
        parts: list[str] = []
        while True:
            ch = self._advance()
            if ch is None:
                raise self._error("unterminated string")
            if ch == '"':
                break
            if ch == "\\":
                esc = self._advance()
                if esc == "n":
                    parts.append("\n")
                elif esc == "t":
                    parts.append("\t")
                elif esc == "r":
                    parts.append("\r")
                elif esc == "\\":
                    parts.append("\\")
                elif esc == '"':
                    parts.append('"')
                elif esc is None:
                    raise self._error("unterminated string escape")
                else:
                    parts.append("\\")
                    parts.append(esc)
            else:
                parts.append(ch)
        return "".join(parts)

    def _parse_bytes(self) -> Value:
        # Consume "b64#"
        for _ in range(4):
            self._advance()
        start = self._pos
        while self._peek() is not None and self._peek() != ";":
            self._advance()
        b64_str = self._text[start : self._pos].strip()
        data = base64.b64decode(b64_str)
        return Value.bytes_(data)

    def _parse_array(self) -> Value:
        self._expect_char("[")
        items: list[Value] = []
        while True:
            self.skip_ws()
            if self._peek() == "]":
                self._advance()
                break
            items.append(self.parse_value())
            self.skip_ws()
            if self._peek() == ",":
                self._advance()
        return Value.array(items)

    def _parse_object(self) -> Value:
        self._expect_char("{")
        entries: list[tuple[str, Value]] = []
        while True:
            self.skip_ws()
            if self._peek() == "}":
                self._advance()
                break
            key = self._parse_key()
            self._expect_char(":")
            value = self.parse_value()
            self._expect_char(";")
            entries.append((key, value))
        return Value.object(entries)

    def _parse_key(self) -> str:
        self.skip_ws()
        if self._peek() == '"':
            return self._parse_quoted_string()
        return self._parse_identifier()

    def _parse_identifier(self) -> str:
        start = self._pos
        ch = self._peek()
        if ch is None or not (ch.isalpha() or ch == "_"):
            raise self._error("expected identifier")
        self._advance()
        while self._peek() is not None:
            ch = self._peek()
            if ch is not None and (ch.isalnum() or ch == "_"):
                self._advance()
            else:
                break
        return self._text[start : self._pos]

    def _skip_type_annotation(self) -> None:
        if self._remaining().startswith("::"):
            self._advance()  # :
            self._advance()  # :
            while self._peek() is not None:
                ch = self._peek()
                if ch is not None and (ch.isalnum() or ch == "_"):
                    self._advance()
                else:
                    break


# ---------------------------------------------------------------------------
# Pretty-printer
# ---------------------------------------------------------------------------


def pretty_print(value: Value, indent: int = 4) -> str:
    """Pretty-print a Value in canonical Surp text notation.

    >>> pretty_print(Value.uint(42))
    '42'
    """
    parts: list[str] = []
    _write_value(parts, value, indent, 0)
    return "".join(parts)


def _write_value(out: list[str], value: Value, indent: int, depth: int) -> None:
    ind = " " * (indent * depth)
    inner = " " * (indent * (depth + 1))

    vt = value.type

    if vt == ValueType.NULL:
        out.append("null")
    elif vt == ValueType.BOOL:
        out.append("true" if value.data else "false")
    elif vt == ValueType.UINT:
        out.append(str(value.data))
    elif vt == ValueType.INT:
        out.append(str(value.data))
    elif vt == ValueType.FLOAT:
        s = repr(value.data)
        # Ensure always has decimal point
        if "." not in s and "e" not in s and "E" not in s and "inf" not in s.lower() and "nan" not in s.lower():
            s += ".0"
        out.append(s)
    elif vt == ValueType.STR:
        out.append('"')
        for ch in value.data:
            if ch == '"':
                out.append('\\"')
            elif ch == "\\":
                out.append("\\\\")
            elif ch == "\n":
                out.append("\\n")
            elif ch == "\r":
                out.append("\\r")
            elif ch == "\t":
                out.append("\\t")
            else:
                out.append(ch)
        out.append('"')
    elif vt == ValueType.BYTES:
        out.append("b64#")
        out.append(base64.b64encode(value.data).decode("ascii"))
    elif vt == ValueType.ARRAY:
        items = value.data
        if not items:
            out.append("[]")
        elif _is_simple_array(items):
            out.append("[")
            for i, item in enumerate(items):
                if i > 0:
                    out.append(", ")
                _write_value(out, item, indent, depth)
            out.append("]")
        else:
            out.append("[\n")
            for i, item in enumerate(items):
                out.append(inner)
                _write_value(out, item, indent, depth + 1)
                if i < len(items) - 1:
                    out.append(",")
                out.append("\n")
            out.append(ind)
            out.append("]")
    elif vt == ValueType.OBJECT:
        entries = value.data
        if not entries:
            out.append("{}")
        else:
            out.append("{\n")
            for key, val in entries:
                out.append(inner)
                if _is_valid_identifier(key):
                    out.append(key)
                else:
                    out.append('"')
                    out.append(key)
                    out.append('"')
                out.append(": ")
                _write_value(out, val, indent, depth + 1)
                out.append(";\n")
            out.append(ind)
            out.append("}")


def _is_simple_array(items: list[Value]) -> bool:
    return len(items) <= 8 and all(
        v.type
        in (
            ValueType.NULL,
            ValueType.BOOL,
            ValueType.UINT,
            ValueType.INT,
            ValueType.FLOAT,
            ValueType.STR,
        )
        for v in items
    )


def _is_valid_identifier(s: str) -> bool:
    if not s:
        return False
    if not (s[0].isalpha() or s[0] == "_"):
        return False
    return all(c.isalnum() or c == "_" for c in s[1:])
